File: 백준/Silver/1388. 바닥 장식/바닥 장식.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

// https://www.acmicpc.net/problem/1388
public class Main {
    static final int H=0, W=1;
    static int N, M;
    static char[][] map;
    static boolean[][] visited;
    static int[] dr = {1, 0};
    static int[] dc = {0, 1};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        map = new char[N][M];
        visited = new boolean[N][M];

        for(int i=0; i<N; i++) {
            map[i] = br.readLine().toCharArray();
        }//for
        br.close();

        int cnt = 0;
        for(int r=0; r<N; r++) {
            for(int c=0; c<M; c++) {
                if(visited[r][c]) continue;

                visited[r][c] = true;
                if(map[r][c] == '-') dfs(r, c, W, map[r][c]);
                else dfs(r, c, H, map[r][c]);
                cnt++;
            }//for c
        }//for r

        System.out.print(cnt);
    }//main

    private static void dfs(int r, int c, int d, char pattern) {
        int nr = r + dr[d];
        int nc = c + dc[d];
        if(check(nr, nc)|| map[nr][nc] != pattern) return;

        visited[nr][nc] = true;
        dfs(nr, nc, d, pattern);
    }//dfs

    private static boolean check(int r, int c) {
        return r >= N || c >= M || visited[r][c] ;
    }//check

}//class

File: 백준/Silver/1388. 바닥 장식/바닥 장식.kt
// https://www.acmicpc.net/problem/1388

fun main() = with(System.`in`.bufferedReader()) {
    // 세로 크기N과 가로 크기 M
    val (n, m) = readLine()!!
        .split(' ').map { it.toInt() }
    // 기훈이 방
    val map: Array<CharArray> = Array(n) {
        readLine().trim().toCharArray()
    }
    val visited = Array(n) {
        BooleanArray(m)
    }

    val dr = intArrayOf(1, 0)
    val dc = intArrayOf(0, 1)
    val h = 0
    val w = 1
    var totalCnt = 0

    fun dfs(r: Int, c: Int, d: Int) {
        val nr = r + dr[d]
        val nc = c + dc[d]
        if(check(n, m, nr, nc)) return
        if(visited[nr][nc] || map[r][c] != map[nr][nc]) return

        visited[nr][nc] = true
        dfs(nr, nc, d)
    }

    var d = h
    for (r in 0 until n) {
        for(c in 0 until m) {
            if (visited[r][c]) continue

            d = if (map[r][c] == '-') w
            else h

            visited[r][c] = true
            dfs(r, c, d)
            totalCnt++
        }
    }

    print(totalCnt)
    close()
}

fun check(n: Int, m: Int, r: Int, c: Int): Boolean {
    return r >= n || c >= m
}
